### PR TITLE
[WIP] Add support for W&B experiment management

### PR DESCRIPTION
**Pull request by AyushExel**
This PR adds:
* Metric and media logging on W&B dashboard. [Here's](https://wandb.ai/cayush/hifigan/runs/1iowf6we?workspace=user-cayush) how the training dashboard for hifigan looks like

TODO:
* Figure out ways to parse W&B args without polluting the API
* Add support for model checkpointing and versioning 
* Add support for dataset visualization

@erogol I'm putting up this draft PR here as I want to discuss how you'd like to handle the W&B-specific arguments. I think the `TrainingArgs` class would be a suitable place to add extra arguments? I haven't worked with Coqpit so I'm not too sure. 
Here are the args that will provide a better experience for users using W&B:
*  allow passing project/ experiment name explicitly
* entity name for collaboration across teams
* arg like `--model_log_period` to log checkpoint of models at specified intervals (and to be able to resume runs directly from these checkpoints)
* arg like `--log_dataset` to log and visualize the dataset

### CODE REVIEW

**Comment by erogol** [comment]
In my head, we just need to replicate `TensorboardLogger and in `BaseTrainingConfig`, we select the logger. The rest is already set by different fields in the config. Does it sound ok?

**Comment by AyushExel** [comment]
@erogol sounds good. I added a few args in `BaseTrainingConfig` and I think that's all we'll need. Apart from dataset visualization all the other things are in place. Just running some tests :)

File: TTS/trainer.py
Comment (erogol) [inline TTS/trainer.py:1116]: It doesn't need to return wandb_logger additionally since I don't think using TB and WDB together is not a common case. Instead we can ask which dashboard that they want to use and just return that. 

Something like 

```python
# in config
log_dashboard: str = "tensorboard" or "wanddb""

# in process_args
if config.log_dashboard == "tensorboard":
    dashboard_logger = TensorboradLogger()
elif config.log_dash....
    dashboard_logger = WandDBLogger()
```

Does it make sense?

File: TTS/config/shared_configs.py
Comment (erogol) [inline TTS/config/shared_configs.py:214]: We can take the name from `run_name`above

File: TTS/config/shared_configs.py
Comment (erogol) [inline TTS/config/shared_configs.py:216]: What does it do in practice?

File: TTS/config/shared_configs.py
Comment (erogol) [inline TTS/config/shared_configs.py:218]: How is this different than `save_step` ?

File: TTS/trainer.py
Comment (erogol) [inline TTS/trainer.py:120]: As below for `process_args`, I think wanddb_logger needs to replace tb_logger if it is the choosen `log_dashboard`. I guess it requires wandb_logger to implement the same function names as tb_logger

File: TTS/trainer.py
Comment (AyushExel) [reply TTS/trainer.py:1116]: Yes it makes sense. I'll make the changes

File: TTS/config/shared_configs.py
Comment (AyushExel) [reply TTS/config/shared_configs.py:214]: The project name is the name of W&B project that is created. One project can contain many runs. For naming the runs, `run_name` from above is used. The project name, if not provided will default to `config.model`. If it makes more sense to just remove this arg and force the project name to be `config.model`, we can do that too.

File: TTS/config/shared_configs.py
Comment (AyushExel) [reply TTS/config/shared_configs.py:216]: I assume you're referring to `wandb_enitity`. Entity is like a team. If you set an entity, all your runs will be logged in a team's dashboard where the member can view your run, multiple members can log runs in the same project as long as they set the same entity. They can read each other's reports etc. It's meant for collaboration.

File: TTS/config/shared_configs.py
Comment (AyushExel) [reply TTS/config/shared_configs.py:218]: `save_step` is meant for local checkpointing. This is meant to log checkpoints in cloud( you W&B project dashboard. See how it [looks ](https://wandb.ai/cayush/Tacotron2/artifacts/model/rvjlrp7t_checkpoint/da6007b538e933eab163)). We can use `save_step` for both things but I thought it'd be better to let the user decide as logging models take bandwidth. 
We can go with whatever you prefer. Let me know :)

File: TTS/trainer.py
Comment (AyushExel) [reply TTS/trainer.py:120]: Okay, I'll try to do it. We might need one extra condition to flush the metrics after every epoch for wandb_logger. Other than that I think all the things can basically remain the same.

File: TTS/config/shared_configs.py
Comment (erogol) [reply TTS/config/shared_configs.py:214]: Thanks for the explanation. I think it is fine to keep it as it is then.

File: TTS/config/shared_configs.py
Comment (erogol) [reply TTS/config/shared_configs.py:216]: Thanks again :) 

What happens if you don't define it ?

File: TTS/config/shared_configs.py
Comment (erogol) [reply TTS/config/shared_configs.py:218]: I think it is good to keep it but maybe set it to `save_step`if not defined.

File: TTS/config/shared_configs.py
Comment (AyushExel) [reply TTS/config/shared_configs.py:216]: If it is set to `None`, the runs will be logged in your personal account and not the teams account

File: TTS/config/shared_configs.py
Comment (AyushExel) [reply TTS/config/shared_configs.py:218]: Okay

Comment (AyushExel) [reply TTS/config/shared_configs.py:220]: done

**Comment by AyushExel** [comment]
@erogol I have unified both the loggers now. I think there no `tb` or `wandb_` specific parameter except `wandb_entity`. 
I've removed `eval_stats` and `train_stats` dicts form `TBLogger` as I didn't see them getting used anywhere. I might be wrong though.
Can I get another review please :)

**Comment by erogol** [comment]
Thx for all the changes. PR looks good to me. 

Let me try for in a training first before we merge it. But I think we can merge this PR before the next release.

File: TTS/trainer.py
Comment (erogol) [inline TTS/trainer.py:163]: Maybe you can put this part in a function and place it under TTS/utils/logging/__init__.py, as it is being called in multiple places.

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/trainer.py
Comment (AyushExel) [reply TTS/trainer.py:163]: Yes makes sense. I'll do it :)

Comment (AyushExel) [reply TTS/trainer.py:163]: Done

**Review by erogol — APPROVED** [review]

File: TTS/config/shared_configs.py
Comment (erogol) [inline TTS/config/shared_configs.py:220]: I guess we don't need this docstring anymore .

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/config/shared_configs.py
Comment (AyushExel) [reply TTS/config/shared_configs.py:220]: Ahh yes. Thanks for pointing this out. I'll fix it

Comment (AyushExel) [reply TTS/config/shared_configs.py]: `wandb_entity` is sort of useful as we discussed above. But yeah, I've changed `wandb_log_model_step` to `log_model_step`

**Review by erogol — APPROVED** [review]

**Comment by erogol** [comment]
I think the PR is ready 🚀...  Do you have something more to add?

**Comment by AyushExel** [comment]
@erogol Nope. I think we're good for now. There are some nice dataset visualization features that we could add but let's get this basic version in prod first :)

**Comment by erogol** [comment]
Sounds good. Maybe you may want to add a recipe using W&B to be an example?

**Comment by AyushExel** [comment]
@erogol Yeah that's coming too. We'll do an announcement with a simple blog post. I'll put the same example in the recipes section. I'll keep you in the loop

**Comment by AyushExel** [comment]
Hi @erogol. I'd like to publish an example of this integration on our docs. So it'd be nice if you could provide any timeline for getting this merged.

**Comment by erogol** [comment]
@AyushExel I don't have a time set but I want to merge it for the next release. Probably about 2 weeks range.

**Comment by erogol** [comment]
If you need to change something again you might need to force pull due to reverting some of the PRs.

I rebased your  ✨PR✨  with the dev and merge it to dev.

**Comment by AyushExel** [comment]
thanks !

**Pull request merged**
